src/groups/mqb/mqba/mqba_commandrouter.cpp

```diff
@@ -0,0 +1,377 @@
+    if (!isRoutingNeeded()) {
+        countDownLatch();
+    }
+}
+
+CommandRouter::RoutingMode::RoutingMode()
+{
+}
+
+CommandRouter::RoutingMode::~RoutingMode()
+{
+}
+
+CommandRouter::AllPartitionPrimariesRoutingMode::
+    AllPartitionPrimariesRoutingMode()
+{
+}
+
+CommandRouter::RouteMembers
+CommandRouter::AllPartitionPrimariesRoutingMode::getRouteMembers(
+    mqbcmd::InternalResult* result,
+    mqbi::Cluster*          cluster)
+{
+    NodesVector primaryNodes;
+    bool        isSelfPrimary;
+    bool        success;
+
+    cluster->dispatcher()->execute(
+        bdlf::BindUtil::bind(&mqbi::Cluster::getPrimaryNodes,
+                             cluster,
+                             &primaryNodes,
+                             &isSelfPrimary,
+                             result),
+        cluster);
+
+    cluster->dispatcher()->synchronize(cluster);
+
+    return {.d_nodes = primaryNodes, .d_self = isSelfPrimary};
+}
+
+CommandRouter::SinglePartitionPrimaryRoutingMode::
+    SinglePartitionPrimaryRoutingMode(int partitionId)
+: d_partitionId(partitionId)
+{
+}
+
+CommandRouter::RouteMembers
+CommandRouter::SinglePartitionPrimaryRoutingMode::getRouteMembers(
+    mqbcmd::InternalResult* result,
+    mqbi::Cluster*          cluster)
+{
+    mqbnet::ClusterNode* node;
+    bool                 isSelfPrimary;
+    bool                 success;
+
+    cluster->dispatcher()->execute(
+        bdlf::BindUtil::bind(&mqbi::Cluster::getPartitionPrimaryNode,
+                             cluster,
+                             &node,
+                             &isSelfPrimary,
+                             result,
+                             d_partitionId),
+        cluster);
+
+    cluster->dispatcher()->synchronize(cluster);
+
+    NodesVector nodes;
+    if (node) {
+        // Put node into vector to be acceptable for "routeCommand"
+        nodes.push_back(node);
+    }
+
+    return {
+        .d_nodes = nodes,
+        .d_self  = isSelfPrimary,
+    };
+}
+
+CommandRouter::ClusterRoutingMode::ClusterRoutingMode()
+{
+}
+
+CommandRouter::RouteMembers CommandRouter::ClusterRoutingMode::getRouteMembers(
+    mqbcmd::InternalResult* result,
+    mqbi::Cluster*          cluster)
+{
+    typedef mqbnet::Cluster::NodesList NodesList;
+    // collect all nodes in cluster
+    const NodesList& allNodes = cluster->netCluster().nodes();
+
+    NodesVector nodes;
+
+    for (NodesList::const_iterator nit = allNodes.begin();
+         nit != allNodes.end();
+         nit++) {
+        if (cluster->netCluster().selfNode() != *nit) {
+            nodes.push_back(*nit);
+        }
+    }
+
+    return {
+        .d_nodes = nodes,
+        .d_self =
+            true,  // Cluster routing always requires original node to exec.
+    };
+}
+
+bool CommandRouter::route(mqbcmd::InternalResult* result,
+                          mqbi::Cluster*          relevantCluster)
+{
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(d_routingMode);
```

Comment:
Yes, it would fire and `getCommandRoutingMode()` gives a nullptr when the command does not require routing. This function should not run unless the command requires routing.

Reply:
I see you have code like:
```
if (routeCommandManager.isRoutingNeeded()) {
        ...
        shouldSelfExecute = routeCommandManager.route(&cmdResult, cluster);
}
```
so method access if protected

Reply:
Good